tests/src/core/PluginPinConnectorTest.cpp

```diff
@@ -0,0 +1,632 @@
+		Engine::destroy();
+	}
+
+	//! Verifies correct channel counts
+	void ChannelCounts()
+	{
+		using namespace lmms;
+
+		auto model = Model{nullptr};
+
+		// Channel counts should stay zero until known
+		auto pcNxN = PluginPinConnector{DynamicChannelCount, DynamicChannelCount, false, &model};
+		QCOMPARE(pcNxN.in().channelCount(), 0);
+		QCOMPARE(pcNxN.out().channelCount(), 0);
+
+		pcNxN.setPluginChannelCountIn(4);
+		QCOMPARE(pcNxN.in().channelCount(), 4);
+		QCOMPARE(pcNxN.out().channelCount(), 0);
+
+		pcNxN.setPluginChannelCountOut(8);
+		QCOMPARE(pcNxN.in().channelCount(), 4);
+		QCOMPARE(pcNxN.out().channelCount(), 8);
+
+		// stereo/stereo effect
+		auto pc2x2 = PluginPinConnector{2, 2, false, &model};
+		QCOMPARE(pc2x2.in().channelCount(), 2);
+		QCOMPARE(pc2x2.out().channelCount(), 2);
+
+		// stereo instrument
+		auto pc0x2 = PluginPinConnector{0, 2, true, &model};
+		QCOMPARE(pc0x2.in().channelCount(), 0);
+		QCOMPARE(pc0x2.out().channelCount(), 2);
+	}
+
+	//! Verifies that the correct default connections are used for different channel counts
+	void DefaultConnections()
+	{
+		using namespace lmms;
+
+		auto model = Model{nullptr};
+
+		// 2 inputs, 2 outputs (stereo/stereo effect)
+		//
+		// In    Out
+		//  ___   ___
+		// |X| | |X| |
+		// | |X| | |X|
+		//  ---   ---
+
+		auto pc2x2 = PluginPinConnector{2, 2, false, &model};
+		QCOMPARE(pc2x2.in().enabled(0, 0), true);
+		QCOMPARE(pc2x2.in().enabled(0, 1), false);
+		QCOMPARE(pc2x2.in().enabled(1, 0), false);
+		QCOMPARE(pc2x2.in().enabled(1, 1), true);
+
+		QCOMPARE(pc2x2.out().enabled(0, 0), true);
+		QCOMPARE(pc2x2.out().enabled(0, 1), false);
+		QCOMPARE(pc2x2.out().enabled(1, 0), false);
+		QCOMPARE(pc2x2.out().enabled(1, 1), true);
+
+		// 1 input, 1 output (mono/mono effect)
+		//
+		// In    Out
+		//  _     _
+		// |X|   |X|
+		// |X|   |X|
+		//  -     -
+
+		auto pc1x1 = PluginPinConnector{1, 1, false, &model};
+		QCOMPARE(pc1x1.in().enabled(0, 0), true);
+		QCOMPARE(pc1x1.in().enabled(1, 0), true);
+
+		QCOMPARE(pc1x1.out().enabled(0, 0), true);
+		QCOMPARE(pc1x1.out().enabled(1, 0), true);
+
+		// 1 input, >2 outputs
+		//
+		// In    Out
+		//  _     _______
+		// |X|   |X| | | |
+		// |X|   | |X| | |
+		//  -     -------
+
+		auto pc1x4 = PluginPinConnector{1, 4, false, &model};
+		QCOMPARE(pc1x4.in().enabled(0, 0), true);
+		QCOMPARE(pc1x4.in().enabled(1, 0), true);
+
+		QCOMPARE(pc1x4.out().enabled(0, 0), true);
+		QCOMPARE(pc1x4.out().enabled(0, 1), false);
+		QCOMPARE(pc1x4.out().enabled(0, 2), false);
+		QCOMPARE(pc1x4.out().enabled(0, 3), false);
+		QCOMPARE(pc1x4.out().enabled(1, 0), false);
+		QCOMPARE(pc1x4.out().enabled(1, 1), true);
+		QCOMPARE(pc1x4.out().enabled(1, 2), false);
+		QCOMPARE(pc1x4.out().enabled(1, 3), false);
+
+		// 2 inputs, 2 outputs (stereo instrument with stereo sidechain input)
+		//
+		// In    Out
+		//  ___   ___
+		// | | | |X| |
+		// | | | | |X|
+		//  ---   ---
+
+		auto pc2x2Inst = PluginPinConnector{2, 2, true, &model};
```

Comment:
Aside from Audio, and Control ports, Lv2 also offers [Cv Ports](https://lv2plug.in/ns/lv2core#CVPort) (briefly explained, control ports with audio rate). Zyn has none btw. The Lv2 docs say CV ports can be represented like Control ports, e.g. with knobs. However, they also say "It is generally safe to connect an audio output to a CV input, but not vice-versa". At that point, I wonder if I should add pins for the Input of such Lv2 plugins. I guess yes, and I guess for the Pin Connector itself, this question is transparent - my Lv2 plugin would just tell the pin connector to reserve another pin and buffer, the Lv2 code then processes the buffer.